lightning/src/routing/router.rs

```diff
@@ -1512,6 +1516,78 @@ where L::Target: Logger {
 		}
 	}
 
+	// Step (10).
+	// Add a random 'shadow route' offset to the CLTV expiry delta values to improve privacy.
+	for path in selected_paths.iter_mut() {
+		let mut shadow_ctlv_expiry_delta_offset: u32 = 0;
+
+		// Choose the last publicly known node as the starting point for the random walk
+		if let Some(starting_hop) = path.iter().rev().find(|h| network_nodes.contains_key(&NodeId::from_pubkey(&h.as_ref().unwrap().pubkey))) {
+			let mut cur_node_id = NodeId::from_pubkey(&starting_hop.as_ref().unwrap().pubkey);
+
+			// Init PRNG with path nonce
+			let mut path_nonce = [0u8; 8];
+			path_nonce.copy_from_slice(&cur_node_id.as_slice()[..8]);
+			let mut prng = ChaCha20::new(random_seed_bytes, &path_nonce);
+			let mut random_path_bytes = [0u8; 8];
+
+			// Pick a random path length in [1 .. 3]
+			prng.process_in_place(&mut random_path_bytes);
+			let random_walk_length = usize::from_be_bytes(random_path_bytes).wrapping_rem(3).wrapping_add(1);
+
+			for _random_hop in 0..random_walk_length { 
+				if let Some(cur_node) = network_nodes.get(&cur_node_id) {
+					// Randomly choose the next hop
+					prng.process_in_place(&mut random_path_bytes);
+					if let Some(random_index) = usize::from_be_bytes(random_path_bytes).checked_rem(cur_node.channels.len()) {
+						if let Some(random_channel_id) = cur_node.channels.get(random_index) {
+							if let Some(random_channel) = network_channels.get(random_channel_id) {
+								if random_channel.node_one == cur_node_id {
+									cur_node_id = random_channel.node_two;
+									if let Some(dir_info) = &random_channel.one_to_two {
+										shadow_ctlv_expiry_delta_offset = shadow_ctlv_expiry_delta_offset
+											.checked_add(dir_info.cltv_expiry_delta.into()).unwrap_or(shadow_ctlv_expiry_delta_offset);
+									}
+								} else {
+									cur_node_id = random_channel.node_one;
+									if let Some(dir_info) = &random_channel.two_to_one {
+										shadow_ctlv_expiry_delta_offset = shadow_ctlv_expiry_delta_offset
+											.checked_add(dir_info.cltv_expiry_delta.into()).unwrap_or(shadow_ctlv_expiry_delta_offset);
+									}
+								};
+							}
+						}
+					}
+				}
+			}
+		} else {
+			// If the entire path is private, choose a random offset from multiples of 40, i.e., the most used cltv_expiry_delta
+			let mut prng = ChaCha20::new(random_seed_bytes, &[0u8; 8]);
+			let mut random_bytes = [0u8; 4];
+			prng.process_in_place(&mut random_bytes);
+			let random_walk_length = u32::from_be_bytes(random_bytes).wrapping_rem(3).wrapping_add(1);
+			shadow_ctlv_expiry_delta_offset = random_walk_length.wrapping_mul(40);
+		}
+
+		// Limit the offset to reduce the payment failure probability
```

Comment:
I'm not sure if limiting the offset reduce the payment failure, i.e the HTLC not being settled by a preimage delivery. Instead, it does increase the channel closing costs in case of irresponsive counterparty along the path, wasting the liquidity locked along the path. 

Maybe it could say "Limit the offset to reduce the worst-case liquidity timevalue in case of defaulting intermediary hop".

Reply:
Yes, adding a large offset would increase the worst-case liquidity cost. For this reason, intermediaries also enforce upper CLTV delta limits during payment forwarding (e.g., LND seems to cap it at 2016 https://github.com/lightningnetwork/lnd/blob/master/htlcswitch/link.go#L49). If I'm not mistaken, we would therefore also risk a payment failure, if we would add an unreasonable large offset. 

But you are right, the first and central reason for capping the offset should be our own liquidity. I'll see to include a more descriptive comment.